filewatcher/v2/filewatcher.go

```diff
@@ -0,0 +1,567 @@
+func WithPollingInterval(interval time.Duration) Option {
+	return func(o *opts) {
+		o.pollingInterval = interval
+	}
+}
+
+// WithInitialReadInterval sets the interval to keep retrying to open the file
+// when creating a new FileWatcher, when the file was not initially available.
+//
+// Default to DefaultInitialReadInterval.
+func WithInitialReadInterval(interval time.Duration) Option {
+	return func(o *opts) {
+		o.initialReadInterval = interval
+	}
+}
+
+// WithFileSizeLimit sets the soft file size limit, with the hard limit being
+// 10x (see HardLimitMultiplier) of the set soft limit.
+//
+// This is completely ignored when DirParser is used.
+//
+// If the soft limit is violated,
+// the violation will be reported via slog at error level and prometheus
+// counter of limitopen_softlimit_violation_total,
+// but it does not stop the normal parsing process.
+//
+// If the hard limit is violated,
+// The loading of the file will fail immediately.
+//
+// Default to DefaultMaxFileSize.
+func WithFileSizeLimit(limit int64) Option {
+	return func(o *opts) {
+		o.fileSizeLimit = limit
+	}
+}
+
+// New creates a new FileWatcher.
+//
+// If the path is not available at the time of calling,
+// it blocks until the file becomes available, or context is cancelled,
+// whichever comes first.
+//
+// When logger is non-nil, it will be used to log errors,
+// either returned by parser or by the underlying file system watcher.
+// Please note that this does not include errors returned by the first parser
+// call, which will be returned directly.
+func New[T any](ctx context.Context, path string, parser Parser[T], options ...Option) (*Result[T], error) {
+	opt := opts{
+		fsEventsDelay:       DefaultFSEventsDelay,
+		pollingInterval:     DefaultPollingInterval,
+		initialReadInterval: DefaultInitialReadInterval,
+
+		fileSizeLimit: DefaultMaxFileSize,
+	}
+	for _, o := range options {
+		o(&opt)
+	}
+	hardLimit := opt.fileSizeLimit * HardLimitMultiplier
+
+	var data T
+	var mtime time.Time
+	var files []string
+
+	var lastErr error
+	for {
+		select {
+		default:
+		case <-ctx.Done():
+			return nil, fmt.Errorf(
+				"filewatcher: context canceled while waiting for file(s) under %q to load: %w, last err: %w",
+				path,
+				ctx.Err(),
+				lastErr,
+			)
+		}
+
+		var err error
+		data, mtime, files, err = openAndParse(path, parser, opt.fileSizeLimit, hardLimit)
+		if errors.Is(err, fs.ErrNotExist) {
+			lastErr = err
+			time.Sleep(opt.initialReadInterval)
+			continue
+		}
+		if err != nil {
+			return nil, err
+		}
+		break
+	}
+
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+	for _, path := range files {
+		if err := watcher.Add(path); err != nil {
+			return nil, fmt.Errorf(
+				"filewatcher: failed to watch %q: %w",
+				path,
+				err,
+			)
+		}
+	}
+
+	res := new(Result[T])
+	res.data.Store(&atomicData[T]{
+		data:  data,
+		mtime: mtime,
+	})
+	res.ctx, res.cancel = context.WithCancel(context.Background())
```

Comment:
why not use the input ctx?  or at least have a WithBackgroundContext option?  you might not even need your close (unless people _want_ to check its error) of you let the context contol its lifecycle